src/Sylius/Bundle/LocaleBundle/Form/Extension/DateExtension.php

```diff
@@ -0,0 +1,139 @@
+use Sylius\Bundle\LocaleBundle\Templating\Helper\LocaleHelper;
+use Symfony\Component\Form\FormBuilderInterface;
+use Symfony\Component\Form\FormInterface;
+use Symfony\Component\Form\FormView;
+use Symfony\Component\Form\AbstractTypeExtension;
+
+/**
+ * Extension for Symfony's DateType for a localized calendar-aware date.
+ */
+class DateExtension extends AbstractTypeExtension {
+
+    /**
+     * @var LocaleHelper
+     */
+    protected $localeHelper;
+
+    /**
+     * @param LocaleHelper $helper
+     */
+    public function __construct(LocaleHelper $helper)
+    {
+        $this->localeHelper = $helper;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function buildForm(FormBuilderInterface $builder, array $options)
+    {
+        if ($this->localeHelper->getCurrentCalendar() === 'gregorian') {
+            // We only need to alter default behaviour when
+            // we're having a traditional calendar system.
+            return;
+        }
+
+        if ('single_text' === $options['widget']) {
+            // When date is in string format
+            $builder->addViewTransformer(new StringGregorianToCalendarSystemTransformer(
+                'Y-m-d',
+                $this->localeHelper
+            ));
+
+        } elseif ('array' === $options['input']) {
+            // When date is in array (of `year`, `month`, `day`) format
+            $builder->addViewTransformer(new ArrayGregorianToCalendarSystemTransformer(
+                $this->localeHelper
+            ));
+        }
+
+        switch ($options['widget']) {
+
+            // If widget is type of `choice` so we need to translate choice items
+            // stored in `choices` option of the field to their localized version.
+            case 'choice':
+
+                $this->translateChoices($builder, 'year', 'Y');
+                $this->translateChoices($builder, 'month', 'm');
+                $this->translateChoices($builder, 'day', 'd');
+
+                break;
+        }
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function finishView(FormView $view, FormInterface $form, array $options)
+    {
+        if ('single_text' === $options['widget']) {
+            // Native browsers do not support different calendar systems
+            // so we have to use "text" type instead of "date" for the input.
+            if ($this->localeHelper->getCurrentCalendar() !== 'gregorian') {
+                // We only need to alter default behaviour when
+                // we're having a traditional calendar system.
+                $view->vars['type'] = 'text';
+            }
+        }
+    }
+
+    /**
+     * A helper method to translate choices of a date-related field
+     * to their localized version.
+     *
+     * e.g. For a year field it should
+     *      translate 2014 (gregorian) choice to to 1393 (persian)
+     *
+     * @param FormBuilderInterface $builder Form builder instance
```

Comment:
This variable comment is redundant, of course that variable ``FormBuilderInterface $builder`` is the instance of form builder :) ``@param`` lines below have the same problem.